src/config/system.rs

```diff
@@ -4,30 +4,241 @@ pub const WS_SUB_MANAGER_ID: u32 = 2;
+    MetricStorage,
+    StorageRegistry,
+};
+use std::{
+    collections::hash_map::HashMap,
+    convert::Infallible,
+    sync::{
+        Arc,
+        RwLock,
+    },
+    time::Duration,
+};
+use thiserror::Error;
+use tokio::sync::{
+    mpsc::{
+        unbounded_channel,
+        UnboundedReceiver,
+        UnboundedSender,
+    },
+    oneshot,
+    Notify,
+};
+
+
+//#[cfg(feature = "prometheusd")]
+pub type MetricSender = UnboundedSender<RpcMetrics>;
+//#[cfg(feature = "prometheusd")]
+pub type MetricReceiver = UnboundedReceiver<RpcMetrics>;
+
+#[derive(Debug)]
+pub struct RegistryChannel {
+    registry: Lazy<StorageRegistry>,
+    notify: Notify,
+}
+
+// #[cfg(feature = "prometheusd")]
+#[derive(Debug, Error)]
+pub enum MetricsError {
+    #[error(transparent)]
+    WSError(#[from] crate::websocket::error::WsError),
+    #[error(transparent)]
+    RPCError(#[from] crate::rpc::error::RpcError),
+}
+// #[cfg(feature = "prometheusd")]
+impl From<MetricsError> for String {
+    fn from(error: MetricsError) -> Self {
+        error.to_string()
+    }
+}
+// #[cfg(feature = "prometheusd")]
+#[derive(MetricStorage, Clone, Debug)]
+#[metric(subsystem = "rpc")]
+pub struct RpcMetrics {
+    #[metric(labels("path", "method", "status"), help = "Total number of requests")]
+    requests: prometheus::IntCounterVec,
+    #[metric(labels("path", "method"), help = "latency of request")]
+    duration: prometheus::HistogramVec,
+}
+//#[cfg(feature = "prometheusd")]
+impl RpcMetrics {
+    pub fn init(registry: &StorageRegistry) -> Result<&Self, prometheus::Error> {
+        RpcMetrics::instance(registry)
+    }
+    pub fn requests_complete(&self, path: &str, method: &str, status: &u16, dt: Duration) {
+        self.requests
+            .with_label_values(&[path, method, &status.to_string()])
+            .inc();
+        self.duration
+            .with_label_values(&[path, method])
+            .observe(dt.as_millis() as f64)
+    }
+}
 
 #[cfg(feature = "journald")]
 pub fn log_journald(level: u32, message: &str) {
     use systemd::journal;
     journal::print(level, message);
 }
 
+// #[cfg(feature = "prometheusd")]
+#[derive(Debug)]
+pub enum MetricsCommand<'a> {
+    Flush(),
+    Channel(&'a MetricSender, &'a MetricReceiver),
+    PushLatency(&'a RpcMetrics, &'a RegistryChannel, &'a str, &'a str, f64),
+    PushRequest(
+        &'a RpcMetrics,
+        &'a RegistryChannel,
+        &'a str,
+        &'a str,
+        &'a u16,
+        Duration,
+    ),
+    PushError(
+        &'a RpcMetrics,
+        &'a RegistryChannel,
+        &'a str,
+        &'a str,
+        &'a u16,
+        Duration,
+    ),
+    Push(&'a RpcMetrics, &'a RegistryChannel),
+    Pull(&'a RpcMetrics, &'a RegistryChannel),
+}
+
+// #[cfg(feature = "prometheusd")]
+#[derive(Debug)]
+pub enum MetricChannelCommand {
+    AdminMsg(oneshot::Sender<RpcMetrics>),
+    StatsMsg(oneshot::Sender<RpcMetrics>),
+}
+
+//#[cfg(feature = "prometheusd")]
```

Comment:
we can have this be `#[cfg(not(feature = "prometheusd"))]` to only enable it when prometheus is off. the fn can also share `listen_for_metrics` so either get picked at compile time!